src/AzureExtension/DevBox/DevBoxInstance.cs

```diff
@@ -698,117 +730,139 @@ public IAsyncOperation<ComputeSystemOperationResult> DoPinActionAsync(string loc
+                var exitcode = ExitCodeInvalid;
+                var psi = new ProcessStartInfo();
+                psi.UseShellExecute = true;
+                psi.FileName = string.Format(CultureInfo.InvariantCulture, ProtocolPinString, location, pinAction, WorkspaceId, DisplayName, Environment, Username);
 
-            Process? p = Process.Start(psi);
-            if (p != null)
-            {
-                p.Refresh();
-                AllowSetForegroundWindow(p.Id);
+                Process? process = Process.Start(psi);
+                if (process != null)
+                {
+                    process.Refresh();
+                    AllowSetForegroundWindow(process.Id);
 
-                // This signals to the WindowsApp that it has been given foreground rights
-                EventWaitHandle signalForegroundSet = new EventWaitHandle(false, EventResetMode.AutoReset, WindowsAppEventName);
-                signalForegroundSet.Set();
+                    // This signals to the WindowsApp that it has been given foreground rights
+                    EventWaitHandle signalForegroundSet = new EventWaitHandle(false, EventResetMode.AutoReset, WindowsAppEventName);
+                    signalForegroundSet.Set();
 
-                p.WaitForExit();
-                exitcode = p.ExitCode;
-                if (exitcode == ExitCodeSuccess)
-                {
-                    return new ComputeSystemOperationResult();
+                    process.WaitForExit();
+                    exitcode = process.ExitCode;
+                    if (exitcode == ExitCodeSuccess)
+                    {
+                        UpdateStateForUI();
+                        return new ComputeSystemOperationResult();
+                    }
                 }
-            }
 
-            errorString = $"DoPinActionAsync with location {location} and action {pinAction} failed with exitcode: {exitcode}";
-            _log.Error(errorString);
-            return new ComputeSystemOperationResult(new NotSupportedException(), Resources.GetResource(Constants.DevBoxUnableToPerformOperationKey), errorString);
+                errorString = $"DoPinActionAsync with location {location} and action {pinAction} failed with exitcode: {exitcode}";
+                throw new NotSupportedException(errorString);
+            }
+            catch (Exception ex)
+            {
+                _log.Error(ex, $"Unable to perform the pinning action for DevBox: {DisplayName}");
+                UpdateStateForUI();
+                return new ComputeSystemOperationResult(ex, Constants.OperationsDefaultErrorMsg, "Pinning action failed");
+            }
         }).AsAsyncOperation();
     }
 
     public IAsyncOperation<ComputeSystemOperationResult> PinToStartMenuAsync()
     {
+        // Since there is no 'Pinning' state we'll say that the state is saving
+        StateChanged?.Invoke(this, ComputeSystemState.Saving);

```

Comment:
I'm guessing that this Saving state doesn't get displayed to the user, but what's the purpose of setting this state here?

Reply:
Actually it does get displayed. In the video you'll see it in the environment card `3 seconds` into the video. Dev home will show a progress bar when it sees that the compute system is in one of the following states

```
            ComputeSystemState.Starting:
            ComputeSystemState.Saving:
            ComputeSystemState.Stopping:
            ComputeSystemState.Pausing:
            ComputeSystemState.Restarting:
            ComputeSystemState.Creating:
            ComputeSystemState.Deleting:
```
See this method: 
https://github.com/microsoft/devhome/blob/69d182a7d6be9727f2d1c9b66774a4bb1574f3a2/tools/Environments/DevHome.Environments/ViewModels/ComputeSystemViewModel.cs#L398

you'll see where we call `IsComputeSystemStateTransitioning` which would return `true`. Then we will show the progress bar.

When an operation is completed the extension is expected to update the state of the compute system to a non transitioning state. So a state thats not one of those above. When we see that we remove the progress bar

Reply:
I see.  Should we add a Pinning and Unpinning state then?

Reply:
Yea I think once we finish bug fixing and plan for environments 2.0 I'd like to update the ComputeSystemState's to include Pinning, Unpinning, Reverting Checkpoint , Creating Checkpoint and Deleting Checkpoint. I'll create an issue for that on GitHub.